lib/datasource.js

```diff
@@ -538,24 +524,63 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+
+  assert(DataSource.relationTypes.indexOf(r.type) !== -1, 'Invalid relation type: ' + r.type);
+  assert(isValidRelationName(rn), 'Invalid relation name: ' + rn);
+
+  // non polymorphic belongsTo relations should have `model` defined
+  if (!r.polymorphic && r.type === 'belongsTo' && !r.model) {
```

Comment:
yes of course! I plan to add one per validation. Thanks for quoting the one you have in mind.